### PR TITLE
LF-3675 improve biodiversity calculation performance

### PR DESCRIPTION
**Pull request by antsgar**
**Description**

Changes to how we query the data from GBIF. Essentially, instead of querying for all occurrences and then filtering out those ones that match the kind of species we’re interested in, I’ve re-written it so that we make a query for each class/kingdom (plants, birds, mammals, etc) and then filter out how many unique occurrences there are for each of those classes. This will mean that we’re less likely to hit the 100,000 limit from their API. I’ve also limited concurrency so that we only make calls for one specific boundary concurrent. So if we have one boundary, we’ll do 1 chunk of calls, if we have two we’ll make 2 chunks, and so on. This also means that if we do hit that API limit, the user could solve the problem by splitting their boundary into several smaller ones.

Jira link:
https://lite-farm.atlassian.net/browse/LF-3675

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [x] Other (please explain)

Tested manually with boundaries of various sizes.

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [x] I have added or updated language tags for text that's part of the UI
- [x] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

File: packages/api/src/controllers/insightHelpers.js [hunk 1]
```diff
@@ -236,105 +238,100 @@ export const getBiodiversityAPI = async (pointData, countData) => {
...
-    Insecta: 'Insects',
-    Plantae: 'Plants',
-    Amphibia: 'Amphibians',
-    Mammalia: 'Mammals',
+  const filterDictionary = {
+    Birds: { classKey: 212 },
+    Insects: { classKey: 216 },
+    Plants: { kingdomKey: 6 },
+    Amphibians: { classKey: 131 },
+    Mammals: { classKey: 359 },
```
Comment (SayakaOno) [inline packages/api/src/controllers/insightHelpers.js:246]: Where are these keys from?

File: packages/api/src/controllers/insightHelpers.js [hunk 2]
```diff
@@ -236,105 +238,100 @@ export const getBiodiversityAPI = async (pointData, countData) => {
...
     Insects: 0,
     Plants: 0,
     Amphibians: 0,
-    CropVarieties: 0,
+    CropVarieties: parseInt(countData),
     Mammals: 0,
   };
-  speciesCount['CropVarieties'] = parseInt(countData);
+
+  const parsedSpecies = [];
```
Comment (SayakaOno) [inline packages/api/src/controllers/insightHelpers.js:258]: `Set` seems better to me here...

File: packages/api/src/controllers/insightHelpers.js [hunk 3]
```diff
@@ -236,105 +238,100 @@ export const getBiodiversityAPI = async (pointData, countData) => {
...
+
+  const processBiodiversityResults = (results, label) => {
+    results.map((result) => {
```
Comment (SayakaOno) [inline packages/api/src/controllers/insightHelpers.js:261]: Nothing is returned from this function, so I think `forEach` should be used instead.

File: packages/webapp/public/locales/es/translation.json [hunk 4]
```diff
@@ -856,10 +856,15 @@
       "AMPHIBIANS": "Anfibio",
       "BIRDS": "Aves",
       "CROP_VARIETIES": "Variedades de cultivos",
-      "ERROR": {
+      "API_OFFSET_LIMIT_ERROR": {
+        "BODY": "Los límites de su granja son demasiado amplios, no podemos proveerle un reporte de Biodiversidad. Intente crear límites más pequeños.",
+        "PREVIEW": "Unavailable",
```
Comment (SayakaOno) [inline packages/webapp/public/locales/es/translation.json:861]: "PREVIEW" is in English in all files :)

File: packages/webapp/public/locales/pt/translation.json [hunk 5]
```diff
@@ -856,7 +856,12 @@
       "AMPHIBIANS": "Anfíbios",
       "BIRDS": "Aves",
       "CROP_VARIETIES": "Variedades de cultivos",
-      "ERROR": {
+      "API_OFFSET_LIMIT_ERROR": {
+        "BODY": "MISSING",
+        "PREVIEW": "Unavailable",
+        "TITLE": "MISSING"
```
Comment (SayakaOno) [inline packages/webapp/public/locales/pt/translation.json:862]: I think you forgot to add the title "Tinha um problema" as you did for the other languages.

**Review by SayakaOno — COMMENTED** [review]
I went through the code but could not figure out the GBIF API part. Could you share the link to the doc that was useful? I am confused with "limit 300" and count... I will come back and re-review once I understand how the API works!

A few more things...
- For translations, you can run `pnpm i18n` in `webapp` after adding tags (ex. `t('TAG.SUBTAG')`). The script updated in this [PR](https://github.com/LiteFarmOrg/LiteFarm/pull/2853) inserts  "MISSING" tags automatically and also orders translations alphabetically.
  It would be great if you could run it!
- You duplicated "PREVIEW" and "TITLE" for the API error. I always get indecisive if I should duplicate one or try to use the existing one. (I feel like I should duplicate but am hesitant to do it) Do you have a clear rule to make this kind of decision?

File: packages/api/src/controllers/insightHelpers.js (hunk 1 above)
Comment (antsgar) [reply packages/api/src/controllers/insightHelpers.js:246]: Unfortunately I couldn't find anywhere in the docs where the keys are explicitly specified -- I ended up doing several queries to the API directly (e.g. https://api.gbif.org/v1/occurrence/search?taxonKey=1) and figuring out from the responses which class or kingdom keys matched the class or kingdom names we were previously using

File: packages/api/src/controllers/insightHelpers.js (hunk 2 above)
Comment (antsgar) [reply packages/api/src/controllers/insightHelpers.js:258]: Good idea! I'll update this

File: packages/webapp/public/locales/es/translation.json (hunk 4 above)
Comment (antsgar) [reply packages/webapp/public/locales/es/translation.json:861]: I think this key is actually not being used anywhere, maybe I should just go ahead and delete it?

File: packages/webapp/public/locales/pt/translation.json (hunk 5 above)
Comment (antsgar) [reply packages/webapp/public/locales/pt/translation.json:862]: Good call! 🙂

**Comment by antsgar** [comment]
@SayakaOno thank you for the review! I should have linked the GBIF docs in the PR description, here they are https://www.gbif.org/developer/occurrence. If you go to the "Searching" section you'll find the relevant information. In summary requests are paginated and you can query max 300 records per page, and no more than 100,000 records total.

I didn't know we had a script to update the language tags, that's awesome! Thank you for letting me know, I'll run it.

And as for duplicating the title and preview, I think either way is fine. In this case since they are the same copy for both modals, so it'd be probably simpler to not duplicate to avoid having extra conditions in the component.

**Review by SayakaOno — APPROVED** [review]
@antsgar I was hoping there would be examples in the document... great job on figuring things out! Thanks to your comment, I understood the logic too :)
It works great!

**Review by kathyavini — COMMENTED** [review]
I'm very impressed with both the solution @antsgar and the nice use of Set @SayakaOno 👌

It's working well for me for boundaries with counts under 300 (and fixes the percentages too so they add up to 100% -- I didn't see that mentioned in the PR text but it's a good bonus!), and also handles the too large boundaries gracefully.

However I have a farm in a sightings-dense area with a count of 91,545 and `const { results } = JSON.parse(data);` is crashing for me on one of my iterations because the API has returned this at the end of some normal JSON data:

```
Could not write JSON: (was java.lang.ArrayIndexOutOfBoundsException); 
nested exception is com.fasterxml.jackson.databind.JsonMappingException: 
(was java.lang.ArrayIndexOutOfBoundsException) 
(through reference chain: org.gbif.api.model.common.search.SearchResponse["results"]->java.util.ArrayList[19]->org.gbif.api.model.occurrence.Occurrence["eventDate"])
```

Have you seen this error before? I don't know if there is just some bunk data in this particular location, or this is related to the size of the dataset? It doesn't seem size-related from the error message, but my `i` was at 50100 at the point of the crash.

It did bring down the API for me, and if we don't mind bringing down beta I can throw the same boundary up on beta to look at. It's not a very big boundary, just in an eBird-sightings-dense location.

**Comment by antsgar** [comment]
> I'm very impressed with both the solution @antsgar and the nice use of Set @SayakaOno 👌
> 
> It's working well for me for boundaries with counts under 300 (and fixes the percentages too so they add up to 100% -- I didn't see that mentioned in the PR text but it's a good bonus!), and also handles the too large boundaries gracefully.
> 
> However I have a farm in a sightings-dense area with a count of 91,545 and `const { results } = JSON.parse(data);` is crashing for me on one of my iterations because the API has returned this at the end of some normal JSON data:
> 
> ```
> Could not write JSON: (was java.lang.ArrayIndexOutOfBoundsException); 
> nested exception is com.fasterxml.jackson.databind.JsonMappingException: 
> (was java.lang.ArrayIndexOutOfBoundsException) 
> (through reference chain: org.gbif.api.model.common.search.SearchResponse["results"]->java.util.ArrayList[19]->org.gbif.api.model.occurrence.Occurrence["eventDate"])
> ```
> 
> Have you seen this error before? I don't know if there is just some bunk data in this particular location, or this is related to the size of the dataset? It doesn't seem size-related from the error message, but my `i` was at 50100 at the point of the crash.
> 
> It did bring down the API for me, and if we don't mind bringing down beta I can throw the same boundary up on beta to look at. It's not a very big boundary, just in an eBird-sightings-dense location.
> 
> Edit: I should also add that since it's the GBIF API that's messing up I guess it doesn't really matter why its failing as long as we keep our server from crashing on that line 🙂

@kathyavini ooooh interesting catch! I don't think I've seen that one before. Yeah I think it's okay that we try that out in beta, we'd need to figure out if it's related to these changes or not. Can you send the coordinates for your boundary over so I can repro on my side?

**Review by kathyavini — APPROVED** [review]
It took me so many tries to get that weird flaky error from GBIF that I thought I couldn't test your PR 😆 

But eventually it came and was cleanly caught 🙂👍

**Review by SayakaOno — APPROVED** [review]

**Pull request merged**